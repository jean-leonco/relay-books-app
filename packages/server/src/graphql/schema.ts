import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../core/middlewares/applyMiddlewares';

// import SubscriptionType from './subscription/SubscriptionType';
import MutationType from './type/MutationType';
import QueryType from './type/QueryType';

const _schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
  // subscription: SubscriptionType,
});

applyMiddlewares(_schema);

export const schema = _schema;
