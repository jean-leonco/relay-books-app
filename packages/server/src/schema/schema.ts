import { GraphQLSchema } from 'graphql';

import applyMiddlewares from '../graphql/middlewares/applyMiddlewares';

import MutationType from './MutationType';
import QueryType from './QueryType';

const schema = new GraphQLSchema({
  query: QueryType,
  mutation: MutationType,
});

applyMiddlewares(schema);

export default schema;
