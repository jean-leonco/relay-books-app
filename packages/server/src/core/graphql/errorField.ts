import { GraphQLString } from 'graphql';

const errorField = {
  error: {
    type: GraphQLString,
    description: 'Default error field resolver for mutations',
    resolve: ({ error }) => error,
  },
};

export default errorField;
