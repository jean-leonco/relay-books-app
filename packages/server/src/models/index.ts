import mongoose from 'mongoose';

import User, { IUser } from '../modules/user/UserModel';
import SessionToken, { ISessionToken, SESSION_TOKEN_SCOPES } from '../modules/sessionToken/SessionTokenModel';
import Book, { IBook } from '../modules/book/BookModel';
import Review, { IReview } from '../modules/review/ReviewModel';
import Category, { ICategory } from '../modules/category/CategoryModel';
import Reading, { IReading } from '../modules/reading/ReadingModel';

mongoose.Promise = global.Promise;

export {
  User,
  IUser,
  SessionToken,
  ISessionToken,
  SESSION_TOKEN_SCOPES,
  Book,
  IBook,
  Review,
  IReview,
  Category,
  ICategory,
  Reading,
  IReading,
};
