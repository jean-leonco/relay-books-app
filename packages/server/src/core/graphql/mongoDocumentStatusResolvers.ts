import { GraphQLBoolean, GraphQLString } from 'graphql';

export const mongoDocumentStatusResolvers = {
  isActive: {
    type: GraphQLBoolean,
    description: 'The soft delete status.',
    resolve: (obj) => obj.isActive,
  },
  removedAt: {
    type: GraphQLString,
    description: 'The hard delete status.',
    resolve: (obj) => (obj.removedAt ? obj.removedAt.toISOString() : null),
  },
  createdAt: {
    type: GraphQLString,
    description: 'The date that the node was created.',
    resolve: (obj) => (obj.createdAt ? obj.createdAt.toISOString() : null),
  },
  updatedAt: {
    type: GraphQLString,
    description: 'The date that the node was last updated.',
    resolve: (obj) => (obj.updatedAt ? obj.updatedAt.toISOString() : null),
  },
};
