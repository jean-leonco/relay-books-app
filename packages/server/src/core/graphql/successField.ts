import { GraphQLString } from 'graphql';

const successField = {
  success: {
    type: GraphQLString,
    description: 'Default success field resolver for mutations',
    resolve: ({ success }) => success,
  },
};

export default successField;
