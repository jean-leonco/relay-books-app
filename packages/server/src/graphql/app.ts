/* eslint-disable no-console */
import 'isomorphic-fetch';

import { koaPlayground } from 'graphql-playground-middleware';
import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import convert from 'koa-convert';
import cors from '@koa/cors';
import graphqlHttp from 'koa-graphql';
import koaLogger from 'koa-logger';
import Router from '@koa/router';
import idx from 'idx';
import i18next from 'i18next';
import koaI18next from 'koa-i18next';
import NoIntrospection from 'graphql-disable-introspection';
import { GraphQLError } from 'graphql';

import * as graphqlLoaders from '../loader';

import { JWT_KEY } from '../common/config';

import { KoaContextExt } from '../types';

import UserModel, { IUser } from '../modules/user/UserModel';

import { t } from '../locales/helpers';

import appPlatformMiddleware from '../modules/auth/appPlatformMiddleware';
import authMiddleware from '../modules/auth/authMiddleware';

import { getDataloaders } from './helper';
import { schema } from './schema';
import { i18nMiddleware } from './i18n';

const app = new Koa<any, KoaContextExt>();

if (process.env.NODE_ENV === 'production') {
  app.proxy = true;
}
app.keys = [JWT_KEY];

const loaders = graphqlLoaders;

const router = new Router<any, KoaContextExt>();

app.use(bodyParser());

app.on('error', (error) => {
  console.error('Error while answering request', { error });
});

if (process.env.NODE_ENV !== 'test') {
  app.use(koaLogger());
}

app.use(convert(cors({ maxAge: 86400, origin: '*' })));

router.get('/healthz', async (ctx) => {
  try {
    await UserModel.find().lean<IUser>();
    ctx.body = 'Database working';
    ctx.status = 200;
  } catch (err) {
    ctx.body = err.toString();
    ctx.status = 500;
  }
});

if (process.env.NODE_ENV !== 'production') {
  router.all(
    '/playground',
    koaPlayground({
      endpoint: '/graphql',
    }),
  );
}

// Middleware to get dataloaders
app.use(async (ctx, next) => {
  ctx.dataloaders = getDataloaders(loaders);
  await next();
});

app.use(
  koaI18next(i18next, {
    lookupCookie: 'lang', // detecting language in cookie
    order: ['header', 'cookie'],
    next: true,
  }),
);

app.use(i18nMiddleware);

// avoid requests from unknown clients
if (process.env.NODE_ENV !== 'development') {
  app.use(appPlatformMiddleware);
}

// Middleware to get result from authorization token
app.use(authMiddleware);

const graphqlSettingsPerReq = async (request, ctx, koaContext) => {
  const { dataloaders, user } = koaContext;
  const { appplatform, timezone } = request.header;

  if (process.env.NODE_ENV !== 'test') {
    console.info('Handling request', {
      appplatform,
      unauthorized: !user,
      userId: idx(user, (_) => _._id),
    });
  }

  return {
    graphiql: process.env.NODE_ENV !== 'production',
    schema,
    //rootValue: { request: ctx.req },
    validationRules: process.env.NODE_ENV === 'production' ? [NoIntrospection] : [],
    context: {
      user,
      dataloaders,
      appplatform,
      koaContext,
      timezone,
      t,
    },
    formatError: (error: GraphQLError) => {
      if (error.name && error.name === 'BadRequestError') {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }

      if (error.path || error.name !== 'GraphQLError') {
        console.error(error);
      } else {
        console.log(`GraphQLWrongQuery: ${error.message}`);
      }

      console.error('GraphQL Error', { error });

      if (process.env.NODE_ENV !== 'production') {
        return {
          message: error.message,
          locations: error.locations,
          stack: error.stack,
        };
      } else {
        ctx.status = 400;
        ctx.body = 'Bad Request';
        return {
          message: 'Bad Request',
        };
      }
    },
  };
};

const graphqlServer = graphqlHttp(graphqlSettingsPerReq);
router.all('/graphql', graphqlServer);

app.use(router.routes()).use(router.allowedMethods());

export default app;
