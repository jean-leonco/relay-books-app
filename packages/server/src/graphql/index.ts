/* eslint-disable no-console */
import 'core-js/stable';
import { createServer } from 'http';

import { execute, subscribe } from 'graphql';
import { SubscriptionServer } from 'subscriptions-transport-ws';

import { GRAPHQL_PORT, GRAPHQL_HOST } from '../common/config';
import connectDatabase from '../common/database';

import app from './app';
import { schema } from './schema';

type ConnectionParams = {
  authorization?: string;
};

const runServer = async () => {
  try {
    console.log('\n🔗 Connecting to database...');
    await connectDatabase();
  } catch (error) {
    console.error('Could not connect to database', { error });
    throw error;
  }

  const server = createServer(app.callback());

  server.listen(GRAPHQL_PORT, () => {
    console.info(`\n🚀 Server started at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}`);

    if (process.env.NODE_ENV !== 'production') {
      console.info(`\n🎠 GraphQL Playground available at http://${GRAPHQL_HOST}:${GRAPHQL_PORT}/playground\n`);
    }

    SubscriptionServer.create(
      {
        onConnect: async (connectionParams: ConnectionParams) => {
          console.info('Client subscription connected', connectionParams);
          // @TODO - get user from connectionParams?.authorization
          // @TODO - return context with user
        },
        onDisconnect: () => console.info('Client subscription disconnected'),
        execute,
        subscribe,
        schema,
      },
      {
        server,
        path: '/subscriptions',
      },
    );
  });
};

(async () => {
  console.log('\n📡 Server starting...');
  await runServer();
})();
