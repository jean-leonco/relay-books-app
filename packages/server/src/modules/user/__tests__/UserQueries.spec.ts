import { graphql } from 'graphql';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { schema } from '../../../graphql/schema';

import {
  clearDbAndRestartCounters,
  connectMongoose,
  createUser,
  disconnectMongoose,
  getContext,
  gql,
} from '../../../../test/helpers';
import { User } from '../../../models';

beforeAll(connectMongoose);

beforeEach(async () => {
  await clearDbAndRestartCounters();
  await User.createIndexes();
});

afterAll(disconnectMongoose);

describe('UserQueries', () => {
  it('should query an me', async () => {
    const user = await createUser();

    const query = gql`
      query Q {
        me {
          id
          name
          email
        }
      }
    `;

    const variables = {};
    const rootValue = {};
    const context = await getContext({ user });
    const result = await graphql(schema, query, rootValue, context, variables);

    expect(result.errors).toBeUndefined();
    expect(sanitizeTestObject(result)).toMatchSnapshot();
  });
});
