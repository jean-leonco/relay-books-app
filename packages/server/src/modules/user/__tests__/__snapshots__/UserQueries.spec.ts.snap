// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserQueries should query an me 1`] = `
Object {
  "data": Object {
    "me": Object {
      "email": "user1@example.com",
      "id": "FROZEN-ID",
      "name": "User #1",
    },
  },
}
`;
